Library/Homebrew/utils/lock.sh

```diff
@@ -0,0 +1,40 @@
+# Ref: http://wiki.bash-hackers.org/howto/mutex
+
+HOMEBREW_LOCK_DIR="$HOMEBREW_LIBRARY/lock.d"
+HOMWBREW_LOCK_PID_FILE="$HOMEBREW_LOCK_DIR/PID"
+
+lock() {
```

Comment:
I agree with xu-cheng here; just using `ps` will probably have a larger window of possible failure. This current implementation is racy too, because these file I/O operations are not atomic, especially if I/O buffering is involved. But it seems less likely to fail in practice, and would be easier to diagnose failure on.

Maybe we could use `File.flock` to do true atomic locking operations? That could actually eliminate the race condition, and has the bonus that the lock is automatically released when the process exits so less "cleanup" code would be needed.